cpp/tests/groupby/merge_sets_tests.cpp

```diff
@@ -0,0 +1,356 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <tests/groupby/groupby_test_util.hpp>
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/iterator_utilities.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/concatenate.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/table/table_view.hpp>
+
+namespace {
+
+constexpr bool print_all{false};  // For debugging
+constexpr int32_t null{0};        // Mark for null child elements
+
+using vcol_views = std::vector<cudf::column_view>;
+
+auto merge_sets(vcol_views const& keys_cols, vcol_views const& values_cols)
+{
+  // Append all the keys and lists together.
+  auto const keys   = cudf::concatenate(keys_cols);
+  auto const values = cudf::concatenate(values_cols);
+
+  std::vector<cudf::groupby::aggregation_request> requests;
+  requests.emplace_back(cudf::groupby::aggregation_request());
+  requests[0].values = *values;
+  requests[0].aggregations.emplace_back(cudf::make_merge_sets_aggregation());
+
+  auto gb_obj = cudf::groupby::groupby(cudf::table_view({*keys}));
+  auto result = gb_obj.aggregate(requests);
+  return std::make_pair(std::move(result.first->release()[0]),
+                        std::move(result.second[0].results[0]));
+}
+
+auto all_valids() { return cudf::test::iterator_no_null(); }
+
+auto all_nulls() { return cudf::test::iterator_all_nulls(); }
+
+auto null_at(cudf::size_type idx) { return cudf::test::iterator_with_null_at(idx); }
+
+auto null_at(std::vector<cudf::size_type> const& indices)
+{
+  return cudf::test::iterator_with_null_at(cudf::host_span<cudf::size_type const>{indices});
+}
```

Comment:
These are repeated from the other file. Factor out?

Reply:
The original names are so long to use (such as `cudf::test::iterator_no_null()`). I want to have shorter names instead, but it's more difficult to change the existing APIs than aliasing them into local functions with much cleaner and shorter names.

Reply:
That's fine, but two .cpp files in the same folder have the same functions. These two .cpp files end up in the same test binary, I believe, so all you need to do is only declare them in both of the .cpp files, but define them only once.

Reply:
Thanks Mark. Unfortunately, these functions return `thrust::transform_iterator` thus we can't declare them multiple times. Thanks for your suggestion, I have another idea that can help to improve the entire testing suite better (which also benefit our future development). I just pushed a PR for it here: https://github.com/rapidsai/cudf/pull/8540

Reply:
The PR https://github.com/rapidsai/cudf/pull/8540 blocks this one. After it is merged, I will refactor this one to use the new implementation.

Reply:
> Thanks Mark. Unfortunately, these functions return thrust::transform_iterator thus we can't declare them multiple times.

Umm, you can make a header.

Reply:
Done. These repeated functions are now gone :)